### ISSUE TITLE
Should Cryptography reqs be Level 1

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#2477]
We currently have an explanation of the thinking around L1 [which can be found here](https://appsecg.host/level1).

There are a few major areas where we have differing opinions about whether a particular type of control falls into the category of L1 for version 5.0.

This issue is to discuss cryptography requirements from the current [chapter 6](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x14-V6-Cryptography.md).

The only requirements being considered for L1 are below (ignore the level marking below). It is agreed (between Josh and Elar for now) that the others in this chapter are not L1.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **6.3.1** | [GRAMMAR, LEVEL L2 > L1] Verify that all random numbers and strings which are intended to be non-guessable must be generated using a cryptographically-secure pseudo-random number generator (CSPRNG) and have at least 128 bits of entropy. Note that UUIDs do not respect this condition. | ✓ | ✓ | ✓ | 338 |
| **6.5.1** | [ADDED, SPLIT FROM 6.2.5] Verify that insecure block modes (e.g., ECB) and weak padding schemes (e.g., PKCS#1 v1.5) are not used. | | ✓ | ✓ | 326 |
| **6.5.2** | [ADDED, SPLIT FROM 6.2.5, LEVEL L2 > L1] Verify that insecure ciphers, including Triple-DES and Blowfish, are not used but secure ciphers and modes such as AES with GCM are. | ✓ | ✓ | ✓ | 326 |
| **6.6.3** | [ADDED, SPLIT FROM 6.2.5] Verify that cryptographic systems avoid the use of disallowed hash functions, such as MD5, SHA-1, or any other insecure hash functions, for any cryptographic purpose. | ✓ | ✓ | ✓ | |

I personally believe that these requirements **do** belong in L1 for the following reasons:

- Many controls (session identifiers, unique file identifiers, one time passwords, etc) rely on random numbers if random numbers are not being created securely then that potentially makes these controls ineffective.
- Similarly many controls (signed tokens, encrypted storage, file integrity checking) rely on cryptography for encryption, signatures and hash generation and use of insecure algorithms potentially makes these controls ineffective.
- Bad cryptography is something that is very hard to fix after the fact. The effort to implement at an earlier stage will be far less than the effort to fix this at a later date.
- Given the importance of cryptography in cyber security in general, the idea that not a single requirement from this chapter would be L1 does not seem reasonable.

**Comment by tghosth** [comment]
> In general, it violates the idea that L1 should have a low barrier to entry.

I don't believe that using good crypto algorithms has a high barrier to entry. This is a pretty low bar and these requirements involve using very standard and well supported algorithms.

> If I am honest, I am not sure how exactly this would be verified anyway? Certainly it would be problematic to verify "black box" style.

Makes up for it by being quite easy to verify white box (unlike input validation) and in many cases there will be ways of doing this black box as well, e.g. burp sequencer for insufficient random, https://hashes.com/en/tools/hash_identifier.

**Comment by elarlang** [comment]
One this is for sure, we have quite a different view on this. You have you beliefs but I have not seen arguments on attack scenarios.

**Comment by randomstuff** [comment]
I would include these in L1 as well:

> 6.2.2 Verify that industry proven or government approved cryptographic algorithms, modes, and libraries are used, instead of custom coded cryptography.
>
> 6.5.3 Verify that nonces, initialization vectors, and other single-use numbers are not used for more than one encryption key/data-element pair. The method of generation must be appropriate for the algorithm being used.

Reasons:

* 6.2.2 → it's not that difficult to check that you are not including some custom coded cryptography;
* 6.5.3 → reusing nonces can completely screw your encryption (especially with CTR and CFB) (and signatures as well).

For this requirements and some others, having an simple L1 application which complies with the requirement is not very difficult: don't use crypto in your application code, let the frameworks do the stuff for me and you should be fine. Using custom crypto code without make these verification might lead you to make dangerous design choices (when used with broken crypto) and you would better use a simpler design which avoids cryptography.

Typical examples I have seen in the wild:

* the usage of weak (unauthenticated) encryption (think static encryption key in application code with DES) to "protect" some HTTP request parameters (leading to potential arbitrary file read);
* custom-coded encryption based on "let's scramble those bits around".

**Comment by elarlang** [comment]
Let's be clear - do we talk about 1st or 2nd layer of defense here?

Important - I don't say we don't need those requirements, I want to understand, are those worth the L1 spotlight. Is the likelihood/impact close to SQL injection, trivial authentication or authorization bypass?

> 6.2.2 → it's not that difficult to check that you are not including some custom coded cryptography;

There are many things that are not difficult to check, (e.g. cookie attributes), but it does not make it L1. We move away from "testability".

> 6.5.3 → reusing nonces can completely screw your encryption (especially with CTR and CFB) (and signatures as well).

Can you provide me a "L1 worth" attack scenario using this?

First example is more like authorization problem to solve? Is it better solution, when used some other crypto-algorithm instead of _where some simple user authorization would have been fine_?

Second example - to be not required in L1 does not mean you need to make your own crypto. If we could use that view, we  shoul put a lot of things into L1. The question here is, how critical is this problem to be solved **as a first things to do**.

**Comment by randomstuff** [comment]
> First example is more like authorization problem to solve?  Is it better solution, when used some other crypto-algorithm instead of where some simple user authorization would have been fine?

Yes, I agree that these examples might be more a problem of something else is not quite right in the design of the application (such as the authorization should be fixed).

So there are two cases to consider for the broken/fake crypto usage:

* broken/fake crypto is used as a misguided stop-gap measure to fix/hide another real problem in the implementation/design of the application and this problem should already be covered by another requirement anyway;
* or there is a real reason for using crypto in the first place and reusing nonces could be very bad.

> Can you provide me a "L1 worth" attack scenario using this?

A simple case would be for digital signature (which not covered in the requirement but probably should). If the application is generating tokens using ECDSA and reuses nonces, the private key can be recovered and the token can be forged.

I think we could come up with interesting attacks when some state is stored in encrypted token in cookies. I'll try to expand on this.

If 6.5.1 is L1, it would probably make sense to have 6.5.3 and probably 6.2.2 as well.

**Comment by jmanico** [comment]
+1 @randomstuff I back this general direction as well. Nice example, thank you.

**Comment by elarlang** [comment]
Although kudos are given here, I'm not convinced. Maybe we need to wait the initial L1 requirements list from me and Josh to get alignment and feeling of L1 idea, but I can not see an L1 scenario here. It may come because of lack of knowledge on the crypto-field and that is why I ask those questions.

**Comment by danielcuthbert** [comment]
I'm with @tghosth and feel what we are asking isn't rocket science, and we've ignored cryptograpy for far too long. This should be in L1.

**Comment by elarlang** [comment]
Many of you say, that it is level 1, but I have not seen yet argumentation and facts to back it up. Without that it is just a religious attitude.

I'm not against L1, I just want to see the argumentation to prove that and that it is aligned with the rest of L1 decisions.

I want to see it as a independent and 1-layer problem. No pre-conditions, no other mistakes in the application.

**Comment by unprovable** [comment]
Fully agree with @tghosth and @danielcuthbert here - using most libraries or TLS/SSH/IPSEC/etc. properly is reasonably straightforward to both do and check, so making it a L1 requirement promotes these best practices from the ground up.

**Comment by danielcuthbert** [comment]
@elarlang "and facts to back it up" what facts do you need? how can we convince you?

**Comment by elarlang** [comment]
As I just wrote:
> I want to see it as a independent and 1st-layer problem. No pre-conditions, no other mistakes in the application.

But I think we need to wait updated definition of Level1 first, with requirements agreed to be in level 1 and then you can analyze your statements as well - is it aligned with the list or not. At the moment there is a chance that we may have quite a different understanding of level 1.

**Comment by unprovable** [comment]
Perhaps these are some examples:
* Appropriate configuration of `nginx.conf` (nginx)
* Appropriate configuration of `sshd.conf` (ssh server)
* Appropriate configuration of `httpd.conf` (apache2)

**Comment by elarlang** [comment]
I think those are covered by [V9](https://asvs.dev/v5.0.draft/0x17-V9-Communications/).

**Comment by tghosth** [comment]
### Intro

So encryption in transit is covered by V9, we have a couple of L1 requirement there.

This question is more about encryption at rest.

I would remind everyone that we are aiming for L1 to have far fewer requirements and therefore we have to make hard choices. At the same time, we need to have good coverage of the different areas.

### Proposed Requirements

The current requirements being proposed for L1 from V6 are as follows.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **6.3.1** | [GRAMMAR, COVERS 6.3.2, LEVEL L2 > L1] Verify that all random numbers and strings which are intended to be non-guessable must be generated using a cryptographically-secure pseudo-random number generator (CSPRNG) and have at least 128 bits of entropy. Note that UUIDs do not respect this condition. | ✓ | ✓ | ✓ | 338 |
| **6.5.1** | [ADDED, SPLIT FROM 6.2.5, COVERS 6.2.3] Verify that insecure block modes (e.g., ECB) and weak padding schemes (e.g., PKCS#1 v1.5) are not used. | | ✓ | ✓ | 326 |
| **6.5.2** | [ADDED, SPLIT FROM 6.2.5, COVERS 6.2.3, LEVEL L2 > L1] Verify that only secure, authenticated ciphers and modes such as AES with GCM are used. | ✓ | ✓ | ✓ | 326 |
| **6.6.1** | [ADDED, SPLIT FROM 6.2.5, COVERS 6.2.3] Verify that only approved hash functions are used for general cryptographic use cases, including digital signatures, HMAC, KDF, and random bit generation. Disallowed hash functions, such as MD5, must not be used for any cryptographic purpose. | | ✓ | ✓ | |

### Arguments in favour

My key arguments in favour are as follows:

- Many controls (session identifiers, unique file identifiers, one time passwords, etc) rely on random numbers if random numbers are not being created securely then that potentially makes these controls ineffective.
- Similarly many controls (signed tokens, encrypted storage, file integrity checking) rely on cryptography for encryption, signatures and hash generation and use of insecure algorithms potentially makes these controls ineffective.
- Bad cryptography is something that is very hard to fix after the fact. The effort to implement at an earlier stage will be far less than the effort to fix this at a later date.
- Given the importance of cryptography in cyber security in general, the idea that not a single requirement from this chapter would be L1 does not seem reasonable.

### Responses to counter arguments

My responses to counter arguments (which Elar produced based on things I have said elsewhere 😀):

#### Low barrier to entry

> In general, it violates the idea that L1 should have a low barrier to entry.

Using the correct crypto algorithms is simple when first starting, these are mature well supported algorithms.

For organizations which are already established, there will be more effort involved but there are also well defined paths to upgrading crypto algorithms such as: 

1) Start encrypting new data with better algorithm, using versioning/tagging to identify how data was encrypted
2) Use tags to identify how to decrypt data when it is requested
3) Gradually run an async process to rotate encryption of stored data.
4) Once all data has been rotated, remove support for old algorithm.

Give me that process over implementing a CSP any day 🤣🤣🤣

However, we would also like to avoid companies having to go through that process which is why we want "using correct algorithms" to be one of the first things we do because it is simple to do up front and annoying to fix afterwards.

#### Verifiability

> If I am honest, I am not sure how exactly this would be verified anyway? Certainly it would be problematic to verify "black box" style.

I said this about input validation. Verifying that would require being aware of all rules for all places and then checking that all rules had been implemented for all places. 

For these crypto requirements, it is pretty straightforward to search code for encryption logic and bad algorithms, most SAST tools already do it.

Obviously black box is a problem but we are not explicitly mandating that L1 should be testable black box anyway.


### Attack scenarios

In terms of attack scenarios:

* Company rolls their own encrypted session tokens and uses a weak algorithm. Attacker predicts the token and gains arbitrary access to customer accounts.
* Company relies on SHA-1 hashes as an identifier (looking at you git ;) ) but attacker uses collision to falsify.
* Company relies on insecure signing to verify the authenticity of files uploaded into the application and attacker is able to spoof a file.

Whilst these are not exactly "point and click" attacks, they are also well within the realms of possibility and the consequences of them are more catastrophic than many vulnerabilities because encryption is often used as a single layer of defence and bad encryption may lead to a compromise of the entire security

**Comment by tghosth** [comment]
tl;dr we rely on crypto for loads of stuff in security plus is if we don't have anything about crypto at rest in L1 we will look like clowns

**Comment by elarlang** [comment]
I agree that "adding crypto later" is complicated, but it is the same with many other things. I don't agree to define L1 via "implementation order".

**Comment by elarlang** [comment]
One last attempt to explain my error with the L1 here.

Let's take 6.3.1
> Verify that all random numbers and strings which are intended to be non-guessable must be generated using a cryptographically-secure pseudo-random number generator (CSPRNG) and have at least 128 bits of entropy. Note that UUIDs do not respect this condition.

If "non-guessable" string or number is _easy_ to guess, it is a L1 problem. And that is the complicated part - what is "easy" and where to draw the line.

This requirement disallows UUID, but if the "random number" is done with the UUID(v4), than I think the risk is not worth L1.

---

If we talk about L1 and the applications, that rely on crypto (at rest) to satisfy high data protection needs - I don't the starting point is L1 for this.

---

> if we don't have anything about "crypto at rest" in L1 I think we will look unserious

You may look unserious (or clown, as your wrote initially) if you have whatever other problem in the application that is marked for L2 at the moment but say "but we have crypto so you need to take me seriously"

---

In general, this comment is valid https://github.com/OWASP/ASVS/issues/2477#issuecomment-2549239444

To point out more precisely:
> Important - I don't say we don't need those requirements, I want to understand, are those worth the L1 spotlight. Is the likelihood/impact close to SQL injection, trivial authentication or authorization bypass?

**Comment by jmanico** [comment]
> This requirement disallows UUID, but if the "random number" is done with the UUID(v4), than I think the risk is not worth L1.

I think this is a really astute/positive observation.

Let me see if I understand what you are trying to say in my own words.

UUIDv4 is not technically a secure 138 bit random generator but damn it is pretty solid and using it is not a serious violation of random number requirements. So maybe using UUIDv4 is not a l1 violation, but other much worse violations would be.

I like the idea of seeing SQLi as l1 because of how critical and damaging it is.

Am I close to understanding what you are trying to say?

I’m a fan of this entire line of thought, Elar.

**Comment by elarlang** [comment]
> Am I close to understanding what you are trying to say?

Probably yes.

But update - we discussed it with Josh and we kind of lowered the really-really-must-have expectation for L1 and added more things into L1. So most of the requirements discussed here got in.

With 6.3.1 - the level really depend on context. For example, if it is short living random-token, the crypto requirement is not that important.

For L1 - it must not be guessable, but the problem with that is that it is not clear, what it means and how to measure/verify it (without declaring the expectation for crypto).

From using the ASVS point of view, the risk is that application owners start having some findings in reports for in practice secure tokens, that "you have huge vulnerability - not enough entropy". But it really depends on context and usage.

**Comment by tghosth** [comment]
The bottom line is that 6.3.1 won't be L1 but 6.5.1, 6.5.2 and 6.6.1 should be.

Stay tuned for more updates :)

**Comment by randomstuff** [comment]
> The bottom line is that 6.3.1 won't be L1 but 6.5.1, 6.5.2 and 6.6.1 should be. The reason is that for 6.3.1, there is a spectrum of potential uses for random numbers plus we disallow UUIDs which is a big deal.

Doesn't it mean that the bit threshold of 6.3.1 should be *relaxed* (but not entirely dropped) for L1?

> With 6.3.1 - the level really depend on context. For example, if it is short living random-token, the crypto requirement is not that important.

If the short lived password reset token is actually generated as `hash(concat(current_time_in_seconds, user_id))`, it is quite  bad. People sometimes end of using {really broken solutions for this kind of things](https://blog.securityinnovation.com/attackers-love-your-password-reset-feature) :) (Agreed in these examples, the token is not random so it technically does not apply to 6.3.1)

If the token is a random token generated using a cryptographically secure source but is only 32 bit long, it might be quite bad as well. Even of the token is short lived.

(I'm not really arguing against this but merely discussing …)
